---
title: "Significant methylation gene annotation"
subtitle: "RSTR vs. LTBI, full dataset"
author: "Kim Dill-McFarland, kadm@uw.edu"
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: no
date: "version `r format(Sys.time(), '%B %d, %Y')`"
editor_options: 
  chunk_output_type: console
---
# Setup
Load packages

```{r message=FALSE, warning=FALSE}
# Data manipulation and figures
library(tidyverse)
library(BIGpicture)
library(ggrepel)
# Annotation
library(biomaRt)
library(org.Hs.eg.db)
# Enrichment
library(SEARchways)
library(rrvgo)
library(GO.db)
#Print pretty tables to Rmd
library(knitr)
library(kableExtra)
  options(knitr.kable.NA = '')
```

Set seed

```{r}
set.seed(589)
```

# Load data

```{r}
DMP <- read_csv("results/RSTR.DMP.results.csv.gz", 
                col_types = cols(CHR="c")) 
DMR <- read_csv("results/RSTR.DMR.results.cpgs.csv.gz")
```

# Significant genes

List significant genes

```{r}
DMP.fdr <- 0.2
#DMP genes
DMP.signif <- DMP %>% 
  filter(FDR <= DMP.fdr & feature != "IGR") %>% 
  mutate(gene_to_use = ifelse(is.na(gene),gene_HGNC_hg38, gene)) %>% 
  distinct(gene_to_use) %>% 
  mutate(gene_to_use = str_split(gene_to_use, ";")) %>% 
  unnest(gene_to_use) %>%  
  drop_na(gene_to_use) %>% 
  dplyr::rename(hgnc_symbol = gene_to_use) %>% 
  mutate(group="DMP")
#DMR genes
DMR.signif <- DMR %>% 
  filter( annotation.group != "IGR") %>% 
  distinct(DMR_genes) %>% 
  mutate(DMR_genes = str_split(DMR_genes, "/")) %>% 
  unnest(DMR_genes)  %>% 
  drop_na(DMR_genes) %>% 
  dplyr::rename(hgnc_symbol = DMR_genes) %>% 
  mutate(group="DMR")

#Combine
signif.all <- bind_rows(DMP.signif, DMR.signif) %>% 
  dplyr::select(group, hgnc_symbol)

overlap <- intersect(DMP.signif$hgnc_symbol, DMR.signif$hgnc_symbol)
```

DMP and DMR-associated genes

```{r}
venn::venn(list("DMP"=DMP.signif$hgnc_symbol, "DMR"=DMR.signif$hgnc_symbol))
```

Add ENSEMBL, ENTREZ, and location data.

```{r}
#Get reference genome
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl", 
                   host = "https://useast.ensembl.org")
  #Check genome version
  searchDatasets(mart = ensembl, pattern = "hsapiens")
```

```{r}
#Get HGNC symbols and gene positions
gene.POS <- getBM(attributes = c('hgnc_symbol', 'ensembl_gene_id',
                                 'entrezgene_id',
                                 'chromosome_name',
                                'start_position','end_position',
                                'strand'),
      mart = ensembl) %>% 
  #Rename
  dplyr::rename(CHR=chromosome_name) %>% 
  #Calculate max range of positions from multi-annotations
  filter(!grepl("CHR", CHR)) %>% 
  group_by(CHR, hgnc_symbol, ensembl_gene_id, entrezgene_id, 
           strand) %>% 
  summarise(min.start = min(start_position, na.rm=TRUE),
            max.end = max(end_position, na.rm=TRUE), .groups = "drop")

#annotate signif genes
signif.anno <- signif.all %>% 
  left_join(gene.POS, by="hgnc_symbol")
```

## Hypergeometric enrichment

Broad terms enriched in DMP / DMP genes

#### Gene set descriptions

* Hallmark gene sets (H)
    - Hallmark gene sets summarize and represent specific well-defined biological states or processes and display coherent expression. These gene sets were generated by a computational methodology based on identifying overlaps between gene sets in other MSigDB collections and retaining genes that display coordinate expression. 
* Basic gene sets (C5)
    - Gene sets that contain genes annotated by the same GO term. Includes:
    - BP: biological process
    - CC: cellular component
    - MF: molecular function
* Curated gene sets (C2)
    - Gene sets curated from various sources including online pathway databases, the biomedical literature, and knowledge of domain experts. Includes:
    - CP: Canonical pathways
        * BIOCARTA: BioCarta gene sets
        * KEGG: KEGG gene sets
        * PID: PID gene sets
        * REACTOME: Reactome gene sets

### Run enrichment

Convert genes to list object

```{r message=FALSE, warning=FALSE}
gene.list <- list()
#All signif genes
gene.list[["DMP/R"]] <- signif.anno %>% 
   pull(hgnc_symbol) %>% unique()
gene.list[["DMP"]] <- signif.anno %>% 
  filter(group == "DMP") %>% pull(hgnc_symbol) %>% unique()
gene.list[["DMR"]] <- signif.anno %>% 
  filter(group == "DMR") %>% pull(hgnc_symbol) %>% unique()
```

#### Hallmark (H)

```{r h.1, warning=FALSE, message=FALSE, results=FALSE}
enrichH <- BIGprofiler(gene.list, ID = "SYMBOL", 
                       category = "H")
```

#### Basic gene sets (C5)

```{r c5.1, warning=FALSE, message=FALSE, results=FALSE}
enrichC5 <- BIGprofiler(gene.list, ID = "SYMBOL", 
                        category = "C5", subcategory = "GO:BP")
```

### Enrichment summary

Significant enrichments, FDR $\leq$ 0.1 and > 1 gene overlapping with term.

```{r warning=FALSE, echo=FALSE, message=FALSE}
enrich.results <- bind_rows(enrichH, enrichC5)
  
enrich.results %>% 
  filter(FDR<=0.2 & `k/K`>0.04) %>% 
  dplyr::select(gs_cat, group, size_group,
                pathway, size_pathway, group_in_pathway, 
                genes, `k/K`, FDR) %>% 
  arrange(`k/K`,group, gs_cat, FDR) %>%

  kable(col.names = c("Gene set", "Group", "Genes in group",
                      "Description", 
                      "Genes in term","Genes overlap", "Genes",
                      "k/K","FDR")) %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE) %>% 
  collapse_rows(1:3, valign="top")
```

Save results

```{r}
dir.create("results/enrichment", showWarnings = FALSE, recursive = TRUE)
enrich.results %>% 
  mutate(genes=as.character(genes)) %>% 
write_csv(file = "results/enrichment/RSTR.DMPR.enrichment.csv")

save(enrich.results, file = "results/enrichment/RSTR.DMPR.enrichment.RData")
```

### Explore GO terms

```{r include=FALSE}
# extract a named vector of all GO terms
goterms <- as.data.frame(Term(GOTERM)) %>% 
  rownames_to_column("GOID") %>% 
  dplyr::rename(GO_path = `Term(GOTERM)`)

goterms.OI <- enrich.results %>% 
  filter(FDR<=0.2 & `k/K`>0.04 & gs_cat == "C5") %>% 
  #Fix names to match GO db
  ##Some of these don't matter in this version but keeping in case changes occur
  mutate(GO_path = gsub("GOBP_","",pathway),
         GO_path = gsub("_"," ",GO_path),
         GO_path = tolower(GO_path),
         GO_path = gsub("high density","high-density",GO_path),
         GO_path = gsub("protein containing","protein-containing",GO_path),
         GO_path = gsub("protein lipid","protein-lipid",GO_path),
         GO_path = gsub("anterior posterior","anterior/posterior",GO_path),
         GO_path = gsub("wnt","Wnt",GO_path),
         GO_path = gsub("cell cell","cell-cell",GO_path),
         GO_path = gsub("hormone mediated","hormone-mediated",GO_path),
         GO_path = gsub("g protein coupled","G protein-coupled",GO_path),
         GO_path = gsub("adenylate cyclase activating","adenylate cyclase-activating", GO_path),
         GO_path = gsub("adenylate cyclase modulating","adenylate cyclase-modulating", GO_path),
         GO_path = recode(GO_path,
                          "cell-cell signaling by Wnt"="cell-cell signaling by wnt")) %>%
  left_join(goterms)
```

Group like terms based on gene content.

```{r}
simMatrix <- calculateSimMatrix(goterms.OI$GOID,
                                orgdb="org.Hs.eg.db",
                                ont="BP",
                                method="Rel")
scores <- setNames(-log10(goterms.OI$FDR), goterms.OI$GOID)

reducedTerms <- reduceSimMatrix(simMatrix,
                                scores,
                                threshold=0.7,
                                orgdb="org.Hs.eg.db")
```

```{r echo=FALSE}
# Adapted from rrvgo::scatterPlot to label all points
# scatterPlot(simMatrix, reducedTerms)
set.seed(42)
x <- cmdscale(as.matrix(as.dist(1-simMatrix)), eig=TRUE, k=2)

df <- cbind(as.data.frame(x$points),
            reducedTerms[match(rownames(x$points), reducedTerms$go),
                         c("term", "parent", "parentTerm", "size")]) %>% 
  left_join(goterms.OI, by=c("term"="GO_path")) %>% 
  #recode small groups
  mutate(GO_group = recode(parentTerm,
                           "biological process involved in intraspecies interaction between organisms"="Other",
                           "negative regulation of hippo signaling"="Hippo signaling",
                           "high-density lipoprotein particle remodeling"="HDL remodeling",
                           "lipid export from cell"="Lipid export",
                           "rhythmic behavior"="Other",
                           "cochlea development"="Other",
                           "regulation of fatty acid biosynthetic process"="Fatty acid\nbiosynthesis"))

col.vec <- c("Fatty acid\nbiosynthesis"="#56B4E9",
             "HDL remodeling"="#E69F00",
             "Hippo signaling"="#CC79A7",
             "Lipid export"="#D55E00", 
             "Other"="grey60",
             "none"="grey90")

df_lab <- df %>% 
  group_by(GO_group) %>% 
  mutate(V1=mean(V1),V2=mean(V2),
         V1 = case_when(GO_group == "Hippo signaling" ~ V1+0.15,
                         GO_group == "HDL remodeling" ~ V1-0.13,
                         TRUE ~ V1),
         V2 = case_when(GO_group == "Other" ~ V2+0.04,
                        GO_group == "Fatty acid\nbiosynthesis" ~ V2-0.05,
                        GO_group == "Lipid export" ~ V2-0.03,
                        TRUE ~ V2))

p <- df %>% 
  arrange(desc(GO_group)) %>% 
  ggplot(aes(x=V1, y=V2, color=GO_group)) +
  geom_point(aes(size=`k/K`*100), alpha=.5) +
  scale_color_discrete(guide="none") +
  scale_size_continuous(range=c(5, 15), breaks = c(5,10,15)) +
  labs(size="Percent enrichment") +
  theme_bw() +
  theme(axis.text   = element_blank(),
        axis.title  = element_blank(),
        axis.ticks = element_blank(),
        panel.background = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.position = c(0.85, 0.82),
        panel.border = element_rect(colour = "black", fill=NA, size=1.5),
        legend.box.background = element_rect(colour = "black", size=1.5)
        )
# p
p2 <- p + geom_label_repel(aes(label=term),
                   box.padding=grid::unit(1, "lines"), size=3,
                   max.overlaps = Inf)
p3 <- p + geom_text(data=df_lab, aes(label=GO_group)) +
  scale_color_manual(values=col.vec ,
                     guide="none") 
p2 + p3
ggsave("figs/enrichment.png", p3, width=6, height=6)
```

# STRING

```{r message=FALSE, echo=FALSE}
fake.enrichment <- df %>% 
  dplyr::select(GO_group, genes) %>% 
  unnest(genes) %>% 
  distinct() %>% 
  group_by(GO_group) %>% 
  summarise(genes=list(genes)) %>% 
  dplyr::rename(pathway = GO_group) %>% 
  mutate(FDR = 0.1, group_in_pathway = 2, `k/K` = 1) %>% 
  mutate(pathway= sub("\n"," ", pathway))

col.vec2 <- c("Cell projection organization"="#AE9CCD",
             "Cellular component size"="#7FCEB9",
             "Fatty acid biosynthesis"="#AAD9F4",
             "HDL remodeling"="#F2CF7F",
             "Hippo signaling"="#E5BCD3",
             "Lipid export"="#E7A571", 
             "Other"="#B3B3B3",
             "none"="grey90")

map <- map_string(genes = gene.list[["DMR"]])

str.plot <- plot_string(map, enrichment = fake.enrichment, colors = col.vec2)
str.plot
```

# Transcription factors

Lambert SA, Jolma A, Campitelli LF, Das PK, Yin Y, Albu M, Chen X, Taipale J, Hughes TR, Weirauch MT.(2018) The Human Transcription Factors. Cell. 172(4):650-665. doi: 10.1016/j.cell.2018.01.029. <http://humantfs.ccbr.utoronto.ca/download.php>

Are any DMR associated genes transcription factors?

```{r}
tx <- read_csv("data_Methyl_clean/DatabaseExtract_v_1.01.csv")
DMR.long <- DMR %>% 
  mutate(gene_to_use = str_split(DMR_genes, "/")) %>% 
  unnest(gene_to_use) %>% 
  distinct(DMR, gene_to_use) %>% 
  drop_na(gene_to_use)

tx %>% 
  filter(`HGNC symbol` %in% gene.list[["DMR"]] &
           `Is TF?` =="Yes") %>% 
  dplyr::select(`HGNC symbol`, `Is TF?`:`Motif status`) %>% 
  left_join(DMR.long, by=c("HGNC symbol"="gene_to_use")) %>% 
  kable() %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE)
```



# R session

```{r}
sessionInfo()
```

***
