---
title: "Methylation linked to gene expression"
subtitle: "RSTR vs. LTBI, full dataset"
author: "Kim Dill-McFarland, kadm@uw.edu"
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: no
date: "version `r format(Sys.time(), '%B %d, %Y')`"
editor_options: 
  chunk_output_type: console
---
# Setup
Load packages

```{r message=FALSE, warning=FALSE}
# Data manipulation and figures
library(tidyverse)
#lmekin
# library(coxme)
library(kimma)
#correlation
library(Hmisc)
#Print pretty tables to Rmd
library(knitr)
library(kableExtra)
  options(knitr.kable.NA = '')
  knitr::opts_chunk$set(message=FALSE)
```

Set seed

```{r}
set.seed(589)
```

# Load data
#### Enrichment

List genes in significantly enriched gene sets.

```{r warning=FALSE}
attach("results/enrichment/RSTR.DMPR.enrichment.RData")
enrich <- enrich.results %>% 
  dplyr::select(group, gs_cat, gs_subcat, pathway, 
         group_in_pathway, FDR, genes) %>% 
  #Significant
  filter(group_in_pathway > 1 & FDR < 0.2) %>% 
  #unnest genes
  unnest(genes) %>% 
  drop_na(genes)
```

#### Methylation 

List significant DMP/DMR and filter to enrichment genes

```{r warning=FALSE}
#Significant genes
DMP <- read_csv("results/RSTR.DMP.results.csv.gz", 
                col_types = cols(CHR="c")) %>% 
  #significant
  filter(FDR <= 0.2 & feature != "IGR") %>%
  #consolidate gene anno
  mutate(gene_to_use = ifelse(is.na(gene), gene_HGNC_hg38, gene)) %>% 
  select(-gene, -gene_HGNC_hg38) %>% 
  #separate annos
  mutate(gene_to_use = str_split(gene_to_use, ";")) %>% 
  unnest(gene_to_use) %>%  
  drop_na(gene_to_use) 
DMP.OI <- DMP %>% 
  #enrichment genes
  filter(gene_to_use %in% enrich$genes)

DMR <- read_csv("results/RSTR.DMR.results.cpgs.csv.gz") %>% 
  #separate annos
  mutate(gene_to_use = str_split(DMR_genes, "/")) %>% 
  unnest(gene_to_use) %>%  
  drop_na(gene_to_use)
DMR.OI <- DMR %>% 
  #significant
  filter(annotation.group != "IGR" & !is.na(gene_to_use))  %>% 
  #enrichment genes
  filter(gene_to_use %in% enrich$genes)
```

```{r}
#M values
attach("data_Methyl_clean/methyl_data.RData")
#Filter DMP probes
DMP.M <- as.data.frame(RSTR.M.kin) %>% 
  rownames_to_column("probeID") %>% 
  filter(probeID %in% c(DMP.OI$probeID))
#Filter DMR probes and average within DMR
DMR.M <- as.data.frame(RSTR.M.kin) %>% 
  rownames_to_column("probeID") %>% 
  filter(probeID %in% c(DMR.OI$probeID)) %>% 
  full_join(select(DMR.OI, DMR, probeID)) %>% 
  select(-probeID) %>% 
  group_by(DMR) %>% 
  summarise_all(~mean(., na.rm=TRUE))
```

#### RNA-seq

```{r}
# attach("../RNAseq/RSTR_RNAseq_data_combined_uniqueFULLID.RData")
attach("../RNAseq/RSTR_RNAseq_data_for_eQTM.RData")

rna <- as.data.frame(dat.combined.voom$E) %>% 
  rownames_to_column("gene") %>% 
  #Keep methylation significant genes
  filter(gene %in% c(DMP.OI$gene_to_use, DMR.OI$gene_to_use))
```

In total, `r nrow(rna)` of `r length(unique(c(DMP.OI$gene_to_use, DMR.OI$gene_to_use)))` significant methylation enrichment genes are present in RNA-seq gene expression. The following genes will be analyzed.


```{r echo=FALSE}
sort(intersect(unique(c(DMP.OI$gene_to_use, DMR.OI$gene_to_use)), rna$gene))
```

While these are excluded due to missing RNA data.

```{r echo=FALSE}
sort(setdiff(unique(c(DMP.OI$gene_to_use, DMR.OI$gene_to_use)), rna$gene))
```

## Determine overlap

Match methylation and expression samples by FULLIDNO

```{r}
DMP.M.overlap <- DMP.M %>% 
  pivot_longer(-probeID, names_to = "Sample_Name", values_to = "methyl")  %>% 
  left_join(select(meta.kin, Sample_Name, FULLIDNO)) %>% 
  mutate(Sample_Name = as.numeric(Sample_Name)) %>% 
  filter(FULLIDNO %in% dat.combined.voom$targets$FULLIDNO) %>% 
  arrange(FULLIDNO)

DMR.M.overlap <- DMR.M %>% 
  pivot_longer(-DMR, names_to = "Sample_Name", values_to = "methyl")  %>% 
  left_join(select(meta.kin, Sample_Name, FULLIDNO)) %>% 
  mutate(Sample_Name = as.numeric(Sample_Name)) %>% 
  filter(FULLIDNO %in% dat.combined.voom$targets$FULLIDNO) %>% 
  arrange(FULLIDNO)

rna.overlap <- rna %>% 
  #add meta
  pivot_longer(-gene, names_to = "libID", values_to = "expression") %>% 
  left_join(dat.combined.voom$targets, by = "libID") %>% 
  select(gene, libID, expression, RSID, Sample_Name, FULLIDNO, 
           condition, Sample_Group, 
           KCHCA_AGE_YR_CURRENT, M0_KCVSEX, experiment) %>% 
  filter(FULLIDNO %in% DMP.M.overlap$FULLIDNO) %>% 
  arrange(FULLIDNO)

#Filter kinship
kin.overlap <- kin %>% 
  #rows
  rownames_to_column() %>% 
  left_join(select(meta.kin, Sample_Name, FULLIDNO), 
            by=c("rowname"="Sample_Name")) %>% 
  select(-rowname) %>% rename(rowname=FULLIDNO) %>% 
  #columns
  pivot_longer(-rowname) %>% 
  left_join(select(meta.kin, Sample_Name, FULLIDNO), 
            by=c("name"="Sample_Name")) %>% 
  select(-name) %>% rename(name=FULLIDNO) %>% 
  #overlap
  filter(rowname %in% DMP.M.overlap$FULLIDNO & 
           name %in% DMP.M.overlap$FULLIDNO) %>% 
  arrange(rowname, name) %>% 
  pivot_wider() %>% 
  column_to_rownames()

meta.overlap <- meta.kin %>% 
  filter(FULLIDNO %in% rownames(kin.overlap)) %>% 
  arrange(FULLIDNO)
```

```{r}
#check
identical(unique(DMP.M.overlap$FULLIDNO), 
          unique(DMR.M.overlap$FULLIDNO))
identical(unique(DMP.M.overlap$FULLIDNO), 
          unique(rna.overlap$FULLIDNO))
identical(unique(DMP.M.overlap$FULLIDNO), 
          rownames(kin.overlap))
identical(unique(DMP.M.overlap$FULLIDNO), 
          colnames(kin.overlap))
identical(unique(DMP.M.overlap$FULLIDNO), 
          meta.overlap$FULLIDNO)
```

Save.

```{r}
save(DMP.OI, DMP.M.overlap, 
     DMR.OI, DMR.M.overlap, 
     rna.overlap, kin.overlap,
     meta.overlap,
     file="data_Methyl_clean/eQTM.data.RData")
```

# DMP/R and DEG overlap

```{r}
deg <- read_csv("../RNAseq/RSTR.Mtb.model.results.anno.csv") %>% 
  filter(variable == "conditionTB:Sample_GroupRSTR" & FDR < 0.2)

venn::venn(list("DEG"=deg$gene, "DMP"=DMP$gene_to_use, "DMR"=unique(DMR$gene_to_use)))
```

# Linear mixed effects model

Model gene expression against significant methylation DMP and DMR found in significant enrichment terms. In total, `r length(unique(rna.overlap$FULLIDNO))` donors have both methylation and expression data.

`gene expression ~ methylation + age + sex + kinship`

```{r}
DMR.map <- DMR.OI %>% 
  dplyr::select(DMR,gene_to_use) %>% 
  filter(gene_to_use %in% rna.overlap$gene) %>% 
  distinct()

DMR.M.mat <- DMR.M.overlap %>% 
  filter(DMR %in% DMR.map$DMR) %>% 
  pivot_wider(names_from = DMR, values_from = methyl) %>% 
  dplyr::select(-Sample_Name) %>% 
  column_to_rownames("FULLIDNO")

rna.media <- rna.overlap %>% 
  filter(condition == "MEDIA" & gene %in% DMR.map$gene_to_use) %>% 
  dplyr::select(gene, FULLIDNO, expression) %>% 
  pivot_wider(names_from = FULLIDNO, values_from = expression) %>% 
  column_to_rownames("gene")

rna.tb <- rna.overlap %>% 
  filter(condition == "TB" & gene %in% DMR.map$gene_to_use) %>% 
  dplyr::select(gene, FULLIDNO, expression) %>% 
  pivot_wider(names_from = FULLIDNO, values_from = expression) %>% 
  column_to_rownames("gene")
```

```{r eval=FALSE}
eQTM.media <- kmFit_eQTL(dat.snp = DMR.M.mat, dat.map = DMR.map,
                         genotypeID = "DMR", geneID = "gene_to_use",
                         counts = rna.media, meta = meta.overlap, kin = kin.overlap,
                         model = "~genotype + KCHCA_AGE_YR_CURRENT + M0_KCVSEX + (1|FULLIDNO)",
                         run.lmerel = TRUE, metrics = TRUE,
                         patientID = "FULLIDNO", libraryID = "FULLIDNO")


eQTM.tb <- kmFit_eQTL(dat.snp = DMR.M.mat, dat.map = DMR.map,
                      genotypeID = "DMR", geneID = "gene_to_use",
                      counts = rna.tb, meta = meta.overlap, kin = kin.overlap,
                      model = "~genotype + KCHCA_AGE_YR_CURRENT + M0_KCVSEX + (1|FULLIDNO)",
                      run.lmerel = TRUE, metrics = TRUE,
                      patientID = "FULLIDNO", libraryID = "FULLIDNO")
```

```{r eval=FALSE}
save(eQTM.media, eQTM.tb, file="results/RSTR.eQTM.kimma.RData")
```

```{r echo=FALSE}
load("results/RSTR.eQTM.kimma.RData")
```

```{r echo=FALSE}
eQTM.signif <- eQTM.media$lmerel %>% 
  mutate(condition = "MEDIA") %>% 
  bind_rows(mutate(eQTM.tb$lmerel, condition = "TB")) %>% 
  filter(grepl("^DMR", variable)) %>% 
  group_by(condition) %>% 
  mutate(FDR = p.adjust(pval, method="BH")) %>% 
  ungroup()

eQTM.signif %>% 
  arrange(gene,condition) %>% 
  mutate(pval = cell_spec(pval, color = ifelse(pval < 0.05, "red", "black"))) %>% 
  kable(escape=FALSE) %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE) 
```

# Correlation

```{r echo=FALSE}
eQTM.pairs <- eQTM.media$lmerel %>% 
  mutate(condition = "MEDIA") %>% 
  bind_rows(mutate(eQTM.tb$lmerel, condition = "TB")) %>% 
  filter(grepl("^DMR", variable)) %>% 
  inner_join(dplyr::select(eQTM.signif, genotype, gene)) %>% 
  rename(DMR=genotype)

to.plot <- DMR.M.overlap %>% 
  filter(DMR %in% DMR.map$DMR) %>% 
  inner_join(rna.overlap) %>% 
  inner_join(eQTM.pairs) %>% 
  group_by(DMR, gene) %>% 
  mutate(facet.lab = paste0("Media FDR = ", 
                           signif(unique(FDR[condition=="MEDIA"]), digits=2),
                           "\nMtb FDR = ", 
                           signif(unique(FDR[condition=="TB"]), digits=2))) %>% 
  left_join(dplyr::select(meta.overlap,-Sample_Name))
```

```{r echo=FALSE}
library(Hmisc)
DMR.M.overlap.mat <- DMR.M.overlap %>% 
  dplyr::select(-Sample_Name) %>% 
  pivot_wider(names_from = FULLIDNO, values_from = methyl) %>% 
  column_to_rownames("DMR") %>% 
  t()
identical(rownames(DMR.M.overlap.mat), colnames(rna.media))
identical(rownames(DMR.M.overlap.mat), colnames(rna.tb))

corr.media <- rcorr(DMR.M.overlap.mat, t(rna.media), type="pearson")
corr.tb <- rcorr(DMR.M.overlap.mat, t(rna.tb), type="pearson")

R <- as.data.frame(corr.media$r) %>% 
  rownames_to_column("variable") %>% 
  pivot_longer(-variable, names_to = "gene") %>% 
  mutate(condition = "MEDIA")
R <- as.data.frame(corr.tb$r) %>% 
  rownames_to_column("variable") %>% 
  pivot_longer(-variable, names_to = "gene") %>% 
  mutate(condition = "TB") %>% 
  bind_rows(R) %>% 
  inner_join(distinct(eQTM.pairs, gene, variable)) %>% 
  dplyr::select(condition, everything()) %>% 
  rename(DMR=variable, R=value) %>% 
  arrange(condition,gene)


P <- as.data.frame(corr.media$P) %>% 
  rownames_to_column("variable") %>% 
  pivot_longer(-variable, names_to = "gene") %>% 
  mutate(condition = "MEDIA")
P <- as.data.frame(corr.tb$P) %>% 
  rownames_to_column("variable") %>% 
  pivot_longer(-variable, names_to = "gene") %>% 
  mutate(condition = "TB") %>% 
  bind_rows(P) %>% 
  inner_join(distinct(eQTM.pairs, gene, variable)) %>% 
  dplyr::select(condition, everything()) %>% 
  rename(DMR=variable, pval=value) %>% 
  arrange(condition,gene)

corr.result <- full_join(R,P)
corr.result %>% 
  mutate(pval = cell_spec(pval, color = ifelse(pval < 0.2, "red", "black"))) %>% 
  kable(escape=FALSE) %>% 
  kable_styling(bootstrap_options = "striped", full_width = FALSE) 
```

```{r}
write_csv(corr.result, file = "results/eQTM_correlation.csv")
```

# R session

```{r}
sessionInfo()
```

***
