---
title: "Methylation linked to Mtb-RSTR gene expression"
subtitle: "RSTR vs. LTBI, full dataset"
author: "Kim Dill-McFarland, kadm@uw.edu"
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: no
date: "version `r format(Sys.time(), '%B %d, %Y')`"
editor_options: 
  chunk_output_type: console
---

```{r}
library(tidyverse)
# library(limma)
library(SEARchways)
library(patchwork)
```


# Data

Methylation enrichment

```{r}
load("results/enrichment/RSTR.DMPR.enrichment.RData")
# load("../RNAseq/kimma_IFN.RData")
```

```{r}
signif.enrich <- enrich.results %>% 
  filter(FDR <= 0.2 & `k/K`>0.04) %>% 
  #remove other group
  mutate(GO_group = case_when(grepl("INTRASPECIES|RHYTHMIC|COCHLEA",pathway) ~ "Other",
                              grepl("HIPPO", pathway) ~ "Hippo signaling",
                              grepl("LIPID_EXPORT",pathway) ~ "Lipid export",
                              grepl("LIPOPROTEIN|PROTEIN_CONTAINING_COMPLEX|LIPID_COMPLEX",pathway) ~ "HDL remodeling",
                              grepl("FATTY",pathway) ~ "Fatty acid biosynthesis")) %>% 
  filter(GO_group != "Other") %>% 
  select(GO_group, pathway, genes) %>% 
  rename(methyl_genes = genes)
```

RNAseq fold changes

```{r}
#RNAseq data
load("../RNAseq/RSTR_RNAseq_data_combined_clean.RData")

#kinship data
kin <- read_csv("../data_refs/kinship_Hawn_all.csv")

overlap <- intersect(kin$rowname, dat.combined.voom$targets$FULLIDNO)

kin <- kin %>% 
  filter(rowname %in% overlap) %>% 
  dplyr::select(rowname, all_of(overlap)) %>% 
  arrange(rowname) %>% 
  column_to_rownames()
```

```{r echo=FALSE}
load("results/enrichment/MTB.RSTR.gsea.RData")
```

```{r eval=FALSE}
# run model to get estimates
km.model <- kimma::kmFit(
  dat.combined.voom, patientID = "FULLIDNO", kin=kin,
  model="~condition*Sample_Group+KCHCA_AGE_YR_CURRENT+M0_KCVSEX+experiment+(1|FULLIDNO)",
  run_lmerel = TRUE, use_weights = TRUE,
  run_contrast = TRUE, contrast_var = "condition:Sample_Group")
```

```{r}
fc.all <- km.model$lmerel.contrast %>% 
  mutate(group = paste(contrast_lvl,"-", contrast_ref)) %>% 
  select(group, gene, estimate) %>% 
  filter(group %in% c("TB LTBI - MEDIA LTBI",
                      "TB RSTR - MEDIA RSTR",
                      "MEDIA RSTR - MEDIA LTBI",
                      "TB RSTR - TB LTBI"))
```

# GSEA

```{r}
gsea_MTB_c5 <- BIGsea(gene_df = fc.all, 
                      category = "C5", subcategory = "GO:BP",
                      ID = "SYMBOL")

#filter terms 
gsea_MTB_c5_signif <- gsea_MTB_c5 %>% 
  inner_join(signif.enrich) %>% 
  #Recalc FDR
  group_by(group) %>% 
  mutate(FDR = p.adjust(pval, "BH")) %>% 
  ungroup()
```

```{r}
gsea_format2 <- gsea_MTB_c5_signif  %>% 
  mutate(signif.col = case_when(pval < 0.05 ~ "P < 0.05",
                                TRUE  ~ "NS")) %>% 
  mutate(signif.col = fct_relevel(signif.col, "NS", after=Inf)) %>% 
  mutate(pathway = gsub("GOBP_","",pathway),
         pathway = gsub("_"," ",pathway),
         pathway = tolower(pathway),
         pathway = gsub("high density","high-density",pathway),
         pathway = gsub("protein containing","protein-containing",pathway),
         pathway = gsub("protein lipid","protein-lipid",pathway),
         pathway = gsub("hippo","Hippo",pathway))

plot.ls2 <- list()

for(g in unique(sort(gsea_format2$GO_group))){
  plot.temp <- gsea_format2 %>% 
    filter(GO_group ==g) %>% 
    
    ggplot() +
    aes(x=group,y=NES, fill=signif.col) +
    geom_segment(aes(group, xend = group, y = 0, yend = NES)) +
    geom_point(shape=21, size=3, stroke=1) +
    facet_grid(GO_group~pathway, scales = "free") +
    coord_flip() +
    scale_fill_manual(values = c("P < 0.05"="#D55E00","P < 0.1"="#E69F00",
                                 "P < 0.2"="#F0E442", "P < 0.3"="lightblue",
                                 "NS"="grey80"),
                      drop = FALSE) +
    theme_classic() +
    lims(y=c(-1.5,2.5)) +
    labs(fill="Significance", x="MDM stimulation",
         y="Normalized enrichment score") +
    theme(panel.border = element_rect(colour = "black", fill=NA)) +
    geom_hline(yintercept = 0)
  
  if(g != "Hippo signaling") { plot.temp <- plot.temp + theme(legend.position = "none") }
  
  plot.ls2[[g]] <- plot.temp
}

gsea_plot <- wrap_plots(plot.ls2[c(1,2,4,3)]) +
  plot_layout(widths=c(1,3))
gsea_plot

ggsave("MDM_MTB_GSEA.png", gsea_plot, width=13, height=4)
```

### Save

```{r}
save(km.model, gsea_MTB_c5_signif,
     file="results/enrichment/MTB.RSTR.gsea.RData")
```

# R session

```{r}
sessionInfo()
```

***